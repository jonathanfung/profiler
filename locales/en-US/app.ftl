### Localization for the App UI of Profiler

# Naming convention for l10n IDs: "module-ComponentName--string-summary".
# This allows us to minimize the risk of conflicting IDs throughout the app.
# Please sort alphabetically by (module name, component name). For each module,
# keep strings in order of appearance.

