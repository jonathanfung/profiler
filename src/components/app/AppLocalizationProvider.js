/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
// @flow

import { negotiateLanguages } from '@fluent/langneg';
import { FluentBundle, FluentResource } from '@fluent/bundle';
import { LocalizationProvider, ReactLocalization } from '@fluent/react';
import * as React from 'react';

// Store all translations as a simple object which is available
// synchronously and bundled with the rest of the code.

type Props = {|
  children: React.Node,
|};

const RESOURCES = {
  fr: new FluentResource(
    "intro = &mdash; Application Web pour l'analyse des performances de Firefox"
  ),
  'en-US': new FluentResource(
    'intro = &mdash; Web app for Firefox performance analysis'
  ),
  pl: new FluentResource(
    'intro = Aplikacja internetowa do analizy wydajności przeglądarki Firefox'
  ),
};

// A generator function responsible for building the sequence
// of FluentBundle instances in the order of user's language
// preferences.
function* generateBundles(userLocales) {
  // Choose locales that are best for the user.
  const currentLocales = negotiateLanguages(
    userLocales,
    ['fr', 'en-US', 'pl'],
    {
      defaultLocale: 'en-US',
    }
  );

  for (const locale of currentLocales) {
    const bundle = new FluentBundle(locale);
    bundle.addResource(RESOURCES[locale]);
    yield bundle;
  }
}

// The ReactLocalization instance stores and caches the sequence of generated
// bundles. You can store it in your app's state.
const l10n = new ReactLocalization(generateBundles(navigator.languages));

export class AppLocalizationProvider extends React.PureComponent<Props> {
  render() {
    const { children } = this.props;

    return <LocalizationProvider l10n={l10n}>{children}</LocalizationProvider>;
  }
}
