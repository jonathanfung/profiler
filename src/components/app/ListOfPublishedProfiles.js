/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import memoize from 'memoize-immutable';
import classNames from 'classnames';

import { InnerNavigationLink } from 'firefox-profiler/components/shared/InnerNavigationLink';
import { ProfileMetaInfoSummary } from 'firefox-profiler/components/shared/ProfileMetaInfoSummary';
import { ProfileDeleteButton } from './ProfileDeleteButton';

import {
  listAllUploadedProfileInformationFromDb,
  type UploadedProfileInformation,
} from 'firefox-profiler/app-logic/uploaded-profiles-db';
import { formatSeconds } from 'firefox-profiler/utils/format-numbers';

import type { Milliseconds, StartEndRange } from 'firefox-profiler/types/units';

import './ListOfPublishedProfiles.css';

// This component displays all published profile, and makes it possible to load
// them by clicking on them, or delete them.

const ONE_DAY_IN_MS = 24 * 60 * 60 * 1000;
const ONE_YEAR_IN_MS = 365 * ONE_DAY_IN_MS;

// All formats are lazy-initialized at the first use by using the `memoize`
// tool, because the DateTimeFormat constructor is slow.
// Because there's no parameter to these functions there will only ever be
// memoized once and should be fairly efficient. We can replace with a dedicated
// "lazy" function in the future if performance ever becomes a problem.
const dateFormats = {
  thisDay: memoize(
    () =>
      new Intl.DateTimeFormat(undefined, { hour: 'numeric', minute: 'numeric' })
  ),
  thisYear: memoize(
    () =>
      new Intl.DateTimeFormat(undefined, {
        month: 'short',
        day: 'numeric',
        hour: 'numeric',
        minute: 'numeric',
      })
  ),
  ancient: memoize(
    () =>
      new Intl.DateTimeFormat(undefined, {
        year: 'numeric',
        month: 'short',
        day: 'numeric',
      })
  ),
};

function _formatDate(date: Date, nowTimestamp: Milliseconds) {
  const timeDifference = nowTimestamp - +date;
  if (timeDifference < 0 || timeDifference > ONE_YEAR_IN_MS) {
    return dateFormats.ancient().format(date);
  }
  if (timeDifference > ONE_DAY_IN_MS) {
    return dateFormats.thisYear().format(date);
  }
  return dateFormats.thisDay().format(date);
}

function _formatRange(range: StartEndRange): string {
  return formatSeconds(range.end - range.start, 3, 1);
}

type PublishedProfileProps = {|
  +onProfileDelete: () => void,
  +uploadedProfileInformation: UploadedProfileInformation,
  +nowTimestamp: Milliseconds,
  +withActionButtons: boolean,
|};

type PublishedProfileState = {|
  +confirmDialogIsOpen: boolean,
|};

/**
 * This implements one line in the list of published profiles.
 */
class PublishedProfile extends React.PureComponent<
  PublishedProfileProps,
  PublishedProfileState
> {
  state = {
    confirmDialogIsOpen: false,
  };

  onOpenConfirmDialog = () => {
    this.setState({ confirmDialogIsOpen: true });
  };

  onCloseConfirmDialog = () => {
    this.setState({ confirmDialogIsOpen: false });
  };

  onCloseSuccessMessage = () => {
    this.props.onProfileDelete();
  };

  render() {
    const {
      uploadedProfileInformation,
      nowTimestamp,
      withActionButtons,
    } = this.props;
    const { confirmDialogIsOpen } = this.state;

    let { urlPath } = uploadedProfileInformation;
    if (!urlPath.startsWith('/')) {
      urlPath = '/' + urlPath;
    }
    const slicedProfileToken = uploadedProfileInformation.profileToken.slice(
      0,
      6
    );
    const profileName = uploadedProfileInformation.name
      ? uploadedProfileInformation.name
      : `Profile #${slicedProfileToken}`;
    const smallProfileName = uploadedProfileInformation.name
      ? uploadedProfileInformation.name
      : '#' + slicedProfileToken;

    return (
      <li
        className={classNames('publishedProfilesListItem', {
          publishedProfilesListItem_ConfirmDialogIsOpen: confirmDialogIsOpen,
        })}
      >
        <a
          className="publishedProfilesLink"
          href={urlPath}
          title={`Click here to load profile ${smallProfileName}`}
        >
          <div className="publishedProfilesDate">
            {_formatDate(
              uploadedProfileInformation.publishedDate,
              nowTimestamp
            )}
          </div>
          <div className="publishedProfilesInfo">
            <div className="publishedProfilesName">
              <strong>{profileName}</strong> (
              {_formatRange(uploadedProfileInformation.publishedRange)})
            </div>
            <ProfileMetaInfoSummary meta={uploadedProfileInformation.meta} />
          </div>
        </a>
        {withActionButtons ? (
          <div className="publishedProfilesActionButtons">
            {uploadedProfileInformation.jwtToken ? (
              <ProfileDeleteButton
                buttonClassName="publishedProfilesDeleteButton"
                profileName={profileName}
                smallProfileName={smallProfileName}
                jwtToken={uploadedProfileInformation.jwtToken}
                profileToken={uploadedProfileInformation.profileToken}
                onOpenConfirmDialog={this.onOpenConfirmDialog}
                onCloseConfirmDialog={this.onCloseConfirmDialog}
                onCloseSuccessMessage={this.onCloseSuccessMessage}
              />
            ) : (
              <button
                className="publishedProfilesDeleteButton photon-button photon-button-default"
                type="button"
                title="This profile cannot be deleted because we lack the authorization information."
                disabled
              >
                Delete
              </button>
            )}
          </div>
        ) : null}
      </li>
    );
  }
}

type Props = {|
  withActionButtons: boolean,
  limit?: number,
|};

type State = {|
  uploadedProfileInformationList: null | UploadedProfileInformation[],
|};

export class ListOfPublishedProfiles extends PureComponent<Props, State> {
  _isMounted = false;

  state = {
    uploadedProfileInformationList: null,
  };

  _refreshList = async () => {
    const uploadedProfileInformationList = await listAllUploadedProfileInformationFromDb();
    if (this._isMounted) {
      // It isn't ideal to use a setState here, but this is the only way.
      this.setState({
        // We want to display the list with the most recent uploaded profile first.
        uploadedProfileInformationList: uploadedProfileInformationList.reverse(),
      });
    }
  };

  async componentDidMount() {
    this._isMounted = true;
    this._refreshList();
    window.addEventListener('focus', this._refreshList);
  }

  componentWillUnmount() {
    this._isMounted = false;
    window.removeEventListener('focus', this._refreshList);
  }

  onProfileDelete = () => {
    this._refreshList();
  };

  render() {
    const { limit, withActionButtons } = this.props;
    const { uploadedProfileInformationList } = this.state;

    if (!uploadedProfileInformationList) {
      return null;
    }

    if (!uploadedProfileInformationList.length) {
      return (
        <p className="photon-body-30">No profile has been uploaded yet!</p>
      );
    }

    const reducedUploadedProfileInformationList = limit
      ? uploadedProfileInformationList.slice(0, limit)
      : uploadedProfileInformationList;

    const profilesRestCount =
      uploadedProfileInformationList.length -
      reducedUploadedProfileInformationList.length;

    let profileRestLabel;
    if (profilesRestCount > 0) {
      profileRestLabel = (
        <>See and manage all your recordings ({profilesRestCount} more)</>
      );
    } else {
      profileRestLabel =
        uploadedProfileInformationList.length > 1 ? (
          <>Manage these recordings</>
        ) : (
          <>Manage this recording</>
        );
    }

    const nowTimestamp = Date.now();

    return (
      <>
        <ul className="publishedProfilesList">
          {reducedUploadedProfileInformationList.map(
            uploadedProfileInformation => (
              <PublishedProfile
                onProfileDelete={this.onProfileDelete}
                key={uploadedProfileInformation.profileToken}
                uploadedProfileInformation={uploadedProfileInformation}
                nowTimestamp={nowTimestamp}
                withActionButtons={withActionButtons}
              />
            )
          )}
        </ul>
        {withActionButtons ? null : (
          <p>
            <InnerNavigationLink dataSource="uploaded-recordings">
              {profileRestLabel}
            </InnerNavigationLink>
          </p>
        )}
      </>
    );
  }
}
