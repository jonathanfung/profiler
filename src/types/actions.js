/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { CallTree } from '../profile-logic/call-tree';
import type JSZip from 'jszip';
import type {
  Profile,
  Thread,
  ThreadIndex,
  Pid,
  BrowsingContextID,
} from './profile';
import type {
  CallNodePath,
  CallNodeTable,
  GlobalTrack,
  LocalTrack,
  TrackIndex,
  MarkerIndex,
  OriginsTimeline,
  ActiveTabTimeline,
  ThreadsKey,
} from './profile-derived';
import type { FuncToFuncMap } from '../profile-logic/symbolication';
import type { TemporaryError } from '../utils/errors';
import type { Transform, TransformStacksPerThread } from './transforms';
import type { IndexIntoZipFileTable } from '../profile-logic/zip-files';
import type { TabSlug } from '../app-logic/tabs-handling';
import type {
  UrlState,
  UploadState,
  State,
  UploadedProfileInformation,
} from './state';
import type { CssPixels, StartEndRange, Milliseconds } from './units';

export type DataSource =
  | 'none'
  | 'from-file'
  //  This datasource is used to fetch a profile from Firefox. This used to be
  //  handled by an addon, hence the name, but now this is all inside Firefox.
  | 'from-addon'
  // This is an alias for 'from-addon' until we phase that one out. We
  // introduced it when implementing the "delete profile" functionality, because
  // `from-addon` didn't suit this use-case well. In the future we want to
  // completely replace `from-addon` with this one.
  | 'unpublished'
  | 'local'
  | 'public'
  | 'from-url'
  | 'compare'
  | 'uploaded-recordings';

export type TimelineType = 'stack' | 'category';
export type PreviewSelection =
  | {| +hasSelection: false, +isModifying: false |}
  | {|
      +hasSelection: true,
      +isModifying: boolean,
      +selectionStart: number,
      +selectionEnd: number,
    |};

/**
 * The counts for how many tracks are hidden in the timeline.
 */
export type HiddenTrackCount = {|
  +hidden: number,
  +total: number,
|};

/**
 * A TrackReference uniquely identifies a track.
 * Note that TrackIndexes aren't globally unique: they're unique among global
 * tracks, and they're unique among local tracks for a specific Pid.
 */
export type GlobalTrackReference = {|
  +type: 'global',
  +trackIndex: TrackIndex,
|};
export type LocalTrackReference = {|
  +type: 'local',
  +trackIndex: TrackIndex,
  +pid: Pid,
|};

export type TrackReference = GlobalTrackReference | LocalTrackReference;

/**
 * Active tab track references
 * A TrackReference uniquely identifies a track.
 */
export type ActiveTabGlobalTrackReference = {|
  +type: 'global',
  +trackIndex: TrackIndex,
|};
export type ActiveTabResourceTrackReference = {|
  +type: 'resource',
  +trackIndex: TrackIndex,
|};

export type ActiveTabTrackReference =
  | ActiveTabGlobalTrackReference
  | ActiveTabResourceTrackReference;

export type RequestedLib = {|
  +debugName: string,
  +breakpadId: string,
|};
export type ImplementationFilter = 'combined' | 'js' | 'cpp';
// Change the strategy for computing the summarizing information for the call tree.
export type CallTreeSummaryStrategy =
  | 'timing'
  | 'js-allocations'
  | 'native-retained-allocations'
  | 'native-allocations'
  | 'native-deallocations-memory'
  | 'native-deallocations-sites';

/**
 * This type determines what kind of information gets sanitized from published profiles.
 */
export type CheckedSharingOptions = {|
  // The following values are for including more information in a sanitized profile.
  includeHiddenThreads: boolean,
  includeFullTimeRange: boolean,
  includeScreenshots: boolean,
  includeUrls: boolean,
  includeExtension: boolean,
  includePreferenceValues: boolean,
|};

export type RightClickedMarkerInfo = {|
  +threadsKey: ThreadsKey,
  +markerIndex: MarkerIndex,
|};

type ProfileAction =
  | {|
      +type: 'ROUTE_NOT_FOUND',
      +url: string,
    |}
  | {|
      +type: 'ASSIGN_TASK_TRACER_NAMES',
      +addressIndices: number[],
      +symbolNames: string[],
    |}
  | {|
      +type: 'CHANGE_SELECTED_CALL_NODE',
      +threadsKey: ThreadsKey,
      +selectedCallNodePath: CallNodePath,
      +optionalExpandedToCallNodePath: ?CallNodePath,
    |}
  | {|
      +type: 'UPDATE_TRACK_THREAD_HEIGHT',
      +height: CssPixels,
      +threadsKey: ThreadsKey,
    |}
  | {|
      +type: 'CHANGE_RIGHT_CLICKED_CALL_NODE',
      +threadsKey: ThreadsKey,
      +callNodePath: CallNodePath | null,
    |}
  | {|
      +type: 'FOCUS_CALL_TREE',
    |}
  | {|
      +type: 'CHANGE_EXPANDED_CALL_NODES',
      +threadsKey: ThreadsKey,
      +expandedCallNodePaths: Array<CallNodePath>,
    |}
  | {|
      +type: 'CHANGE_SELECTED_MARKER',
      +threadsKey: ThreadsKey,
      +selectedMarker: MarkerIndex | null,
    |}
  | {|
      +type: 'CHANGE_SELECTED_NETWORK_MARKER',
      +threadsKey: ThreadsKey,
      +selectedNetworkMarker: MarkerIndex | null,
    |}
  | {|
      +type: 'CHANGE_RIGHT_CLICKED_MARKER',
      +threadsKey: ThreadsKey,
      +markerIndex: MarkerIndex | null,
    |}
  | {|
      +type: 'UPDATE_PREVIEW_SELECTION',
      +previewSelection: PreviewSelection,
    |}
  | {|
      +type: 'CHANGE_SELECTED_ZIP_FILE',
      +selectedZipFileIndex: IndexIntoZipFileTable | null,
    |}
  | {|
      +type: 'CHANGE_EXPANDED_ZIP_FILES',
      +expandedZipFileIndexes: Array<IndexIntoZipFileTable | null>,
    |}
  | {|
      +type: 'CHANGE_GLOBAL_TRACK_ORDER',
      +globalTrackOrder: TrackIndex[],
    |}
  | {|
      +type: 'HIDE_GLOBAL_TRACK',
      +trackIndex: TrackIndex,
      +selectedThreadIndexes: Set<ThreadIndex>,
    |}
  | {|
      +type: 'SHOW_GLOBAL_TRACK',
      +trackIndex: TrackIndex,
    |}
  | {|
      +type: 'SHOW_ALL_TRACKS',
    |}
  | {|
      // Isolate only the process track, and not the local tracks.
      +type: 'ISOLATE_PROCESS',
      +hiddenGlobalTracks: Set<TrackIndex>,
      +isolatedTrackIndex: TrackIndex,
      +selectedThreadIndexes: Set<ThreadIndex>,
    |}
  | {|
      // Isolate the process track, and hide the local tracks.
      type: 'ISOLATE_PROCESS_MAIN_THREAD',
      pid: Pid,
      hiddenGlobalTracks: Set<TrackIndex>,
      isolatedTrackIndex: TrackIndex,
      selectedThreadIndexes: Set<ThreadIndex>,
      hiddenLocalTracks: Set<TrackIndex>,
    |}
  | {|
      // Isolate only the screenshot track
      +type: 'ISOLATE_SCREENSHOT_TRACK',
      +hiddenGlobalTracks: Set<TrackIndex>,
    |}
  | {|
      +type: 'CHANGE_LOCAL_TRACK_ORDER',
      +localTrackOrder: TrackIndex[],
      +pid: Pid,
    |}
  | {|
      +type: 'HIDE_LOCAL_TRACK',
      +pid: Pid,
      +trackIndex: TrackIndex,
      +selectedThreadIndexes: Set<ThreadIndex>,
    |}
  | {|
      +type: 'SHOW_LOCAL_TRACK',
      +pid: Pid,
      +trackIndex: TrackIndex,
    |}
  | {|
      +type: 'ISOLATE_LOCAL_TRACK',
      +pid: Pid,
      +hiddenGlobalTracks: Set<TrackIndex>,
      +hiddenLocalTracks: Set<TrackIndex>,
      +selectedThreadIndexes: Set<ThreadIndex>,
    |}
  | {|
      +type: 'SET_CONTEXT_MENU_VISIBILITY',
      +isVisible: boolean,
    |}
  | {|
      +type: 'INCREMENT_PANEL_LAYOUT_GENERATION',
    |}
  | {| +type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' |}
  | {| +type: 'DISMISS_NEWLY_PUBLISHED' |}
  | {|
      +type: 'ENABLE_EVENT_DELAY_TRACKS',
      +localTracksByPid: Map<Pid, LocalTrack[]>,
      +localTrackOrderByPid: Map<Pid, TrackIndex[]>,
    |};

type ReceiveProfileAction =
  | {|
      +type: 'BULK_SYMBOLICATION',
      +symbolicatedThreads: Thread[],
      +oldFuncToNewFuncMaps: Map<ThreadIndex, FuncToFuncMap>,
    |}
  | {|
      +type: 'DONE_SYMBOLICATING',
    |}
  | {|
      +type: 'TEMPORARY_ERROR',
      +error: TemporaryError,
    |}
  | {|
      +type: 'FATAL_ERROR',
      +error: Error,
    |}
  | {|
      +type: 'PROFILE_LOADED',
      +profile: Profile,
      +pathInZipFile: ?string,
      +implementationFilter: ?ImplementationFilter,
      +transformStacks: ?TransformStacksPerThread,
    |}
  | {|
      +type: 'VIEW_FULL_PROFILE',
      +selectedThreadIndexes: Set<ThreadIndex>,
      +globalTracks: GlobalTrack[],
      +globalTrackOrder: TrackIndex[],
      +hiddenGlobalTracks: Set<TrackIndex>,
      +localTracksByPid: Map<Pid, LocalTrack[]>,
      +hiddenLocalTracksByPid: Map<Pid, Set<TrackIndex>>,
      +localTrackOrderByPid: Map<Pid, TrackIndex[]>,
    |}
  | {|
      +type: 'VIEW_ORIGINS_PROFILE',
      +selectedThreadIndexes: Set<ThreadIndex>,
      +originsTimeline: OriginsTimeline,
    |}
  | {|
      +type: 'VIEW_ACTIVE_TAB_PROFILE',
      +selectedThreadIndexes: Set<ThreadIndex>,
      +activeTabTimeline: ActiveTabTimeline,
      +browsingContextID: BrowsingContextID | null,
    |}
  | {|
      +type: 'DATA_RELOAD',
    |}
  | {| +type: 'RECEIVE_ZIP_FILE', +zip: JSZip |}
  | {| +type: 'PROCESS_PROFILE_FROM_ZIP_FILE', +pathInZipFile: string |}
  | {| +type: 'FAILED_TO_PROCESS_PROFILE_FROM_ZIP_FILE', +error: any |}
  | {| +type: 'DISMISS_PROCESS_PROFILE_FROM_ZIP_ERROR' |}
  | {| +type: 'RETURN_TO_ZIP_FILE_LIST' |}
  | {| +type: 'FILE_NOT_FOUND_IN_ZIP_FILE', +pathInZipFile: string |}
  | {| +type: 'REQUESTING_SYMBOL_TABLE', +requestedLib: RequestedLib |}
  | {| +type: 'RECEIVED_SYMBOL_TABLE_REPLY', +requestedLib: RequestedLib |}
  | {| +type: 'START_SYMBOLICATING' |}
  | {| +type: 'WAITING_FOR_PROFILE_FROM_ADDON' |}
  | {| +type: 'WAITING_FOR_PROFILE_FROM_STORE' |}
  | {| +type: 'WAITING_FOR_PROFILE_FROM_URL', +profileUrl: ?string |}
  | {| +type: 'TRIGGER_LOADING_FROM_URL', +profileUrl: string |};

type UrlEnhancerAction =
  | {| +type: 'START_FETCHING_PROFILES' |}
  | {| +type: 'URL_SETUP_DONE' |}
  | {| +type: 'UPDATE_URL_STATE', +newUrlState: UrlState | null |};

type UrlStateAction =
  | {| +type: 'WAITING_FOR_PROFILE_FROM_FILE' |}
  | {|
      +type: 'PROFILE_PUBLISHED',
      +hash: string,
      +profileName: string,
      +prePublishedState: State | null,
    |}
  | {| +type: 'CHANGE_SELECTED_TAB', +selectedTab: TabSlug |}
  | {| +type: 'COMMIT_RANGE', +start: number, +end: number |}
  | {| +type: 'POP_COMMITTED_RANGES', +firstPoppedFilterIndex: number |}
  | {|
      +type: 'CHANGE_SELECTED_THREAD',
      +selectedThreadIndexes: Set<ThreadIndex>,
    |}
  | {|
      +type: 'SELECT_TRACK',
      +selectedThreadIndexes: Set<ThreadIndex>,
      +selectedTab: TabSlug,
    |}
  | {|
      +type: 'CHANGE_RIGHT_CLICKED_TRACK',
      +trackReference: TrackReference | null,
    |}
  | {| +type: 'CHANGE_CALL_TREE_SEARCH_STRING', +searchString: string |}
  | {|
      +type: 'ADD_TRANSFORM_TO_STACK',
      +threadsKey: ThreadsKey,
      +transform: Transform,
      +transformedThread: Thread,
    |}
  | {|
      +type: 'POP_TRANSFORMS_FROM_STACK',
      +threadsKey: ThreadsKey,
      +firstPoppedFilterIndex: number,
    |}
  | {|
      +type: 'CHANGE_TIMELINE_TYPE',
      +timelineType: TimelineType,
    |}
  | {|
      +type: 'CHANGE_IMPLEMENTATION_FILTER',
      +implementation: ImplementationFilter,
      +threadsKey: ThreadsKey,
      +transformedThread: Thread,
      +previousImplementation: ImplementationFilter,
      +implementation: ImplementationFilter,
    |}
  | {|
      type: 'CHANGE_CALL_TREE_SUMMARY_STRATEGY',
      callTreeSummaryStrategy: CallTreeSummaryStrategy,
    |}
  | {|
      +type: 'CHANGE_INVERT_CALLSTACK',
      +invertCallstack: boolean,
      +callTree: CallTree,
      +callNodeTable: CallNodeTable,
      +selectedThreadIndexes: Set<ThreadIndex>,
    |}
  | {|
      +type: 'CHANGE_SHOW_USER_TIMINGS',
      +showUserTimings: boolean,
    |}
  | {|
      +type: 'CHANGE_SHOW_JS_TRACER_SUMMARY',
      +showSummary: boolean,
    |}
  | {| +type: 'CHANGE_MARKER_SEARCH_STRING', +searchString: string |}
  | {| +type: 'CHANGE_NETWORK_SEARCH_STRING', +searchString: string |}
  | {| +type: 'CHANGE_PROFILES_TO_COMPARE', +profiles: string[] |}
  | {| +type: 'CHANGE_PROFILE_NAME', +profileName: string | null |}
  | {|
      +type: 'SANITIZED_PROFILE_PUBLISHED',
      +hash: string,
      +committedRanges: StartEndRange[] | null,
      +oldThreadIndexToNew: Map<ThreadIndex, ThreadIndex> | null,
      +profileName: string,
      +prePublishedState: State | null,
    |}
  | {|
      +type: 'SET_DATA_SOURCE',
      +dataSource: DataSource,
    |}
  | {|
      +type: 'CHANGE_MOUSE_TIME_POSITION',
      +mouseTimePosition: Milliseconds | null,
    |}
  | {|
      +type: 'TOGGLE_RESOURCES_PANEL',
      +selectedThreadIndexes: Set<ThreadIndex>,
    |}
  | {|
      +type: 'PROFILE_REMOTELY_DELETED',
    |};

type IconsAction =
  | {| +type: 'ICON_HAS_LOADED', +icon: string |}
  | {| +type: 'ICON_IN_ERROR', +icon: string |};

type SidebarAction = {|
  +type: 'CHANGE_SIDEBAR_OPEN_STATE',
  +tab: TabSlug,
  +isOpen: boolean,
|};

type PublishAction =
  | {|
      +type: 'TOGGLE_CHECKED_SHARING_OPTION',
      +slug: $Keys<CheckedSharingOptions>,
    |}
  | {|
      +type: 'UPLOAD_STARTED',
    |}
  | {|
      +type: 'UPDATE_UPLOAD_PROGRESS',
      +uploadProgress: number,
    |}
  | {|
      +type: 'UPLOAD_FAILED',
      +error: mixed,
    |}
  | {|
      +type: 'UPLOAD_ABORTED',
    |}
  | {|
      +type: 'UPLOAD_RESET',
    |}
  | {|
      +type: 'UPLOAD_COMPRESSION_STARTED',
      +abortFunction: () => void,
    |}
  | {|
      +type: 'CHANGE_UPLOAD_STATE',
      +changes: $Shape<UploadState>,
    |}
  | {|
      +type: 'REVERT_TO_PRE_PUBLISHED_STATE',
      +prePublishedState: State,
    |}
  | {| +type: 'HIDE_STALE_PROFILE' |};

type DragAndDropAction =
  | {|
      +type: 'START_DRAGGING',
    |}
  | {|
      +type: 'STOP_DRAGGING',
    |}
  | {|
      +type: 'REGISTER_DRAG_AND_DROP_OVERLAY',
    |}
  | {|
      +type: 'UNREGISTER_DRAG_AND_DROP_OVERLAY',
    |};

type CurrentProfileUploadedInformationAction = {|
  +type: 'SET_CURRENT_PROFILE_UPLOADED_INFORMATION',
  +uploadedProfileInformation: UploadedProfileInformation | null,
|};

export type Action =
  | ProfileAction
  | ReceiveProfileAction
  | SidebarAction
  | UrlEnhancerAction
  | UrlStateAction
  | IconsAction
  | PublishAction
  | DragAndDropAction
  | CurrentProfileUploadedInformationAction;
