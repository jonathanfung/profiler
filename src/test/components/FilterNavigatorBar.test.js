/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import { ProfileFilterNavigator } from '../../components/app/ProfileFilterNavigator';
import * as ProfileView from '../../actions/profile-view';
import * as ReceiveProfile from '../../actions/receive-profile';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/processed-profile';

describe('app/ProfileFilterNavigator', () => {
  const browsingContextID = 123123;
  function setup() {
    const { profile } = getProfileFromTextSamples(`
      A  A  A
      B  B  B
      C  C  H
      D  F  I
      E  G
      `);
    // Add page for active tab.
    profile.pages = [
      {
        browsingContextID: browsingContextID,
        innerWindowID: 1,
        url: 'https://developer.mozilla.org/en-US/',
        embedderInnerWindowID: 0,
      },
    ];
    profile.meta.configuration = {
      threads: [],
      features: [],
      capacity: 1000000,
      activeBrowsingContextID: browsingContextID,
    };

    // Change the root range for testing.
    const samples = profile.threads[0].samples;
    samples.time[samples.length - 1] = 50;

    const store = storeWithProfile(profile);
    const renderResult = render(
      <Provider store={store}>
        <ProfileFilterNavigator />
      </Provider>
    );

    return {
      ...store,
      ...renderResult,
    };
  }

  it('renders ProfileFilterNavigator properly', () => {
    const { container, dispatch } = setup();
    // Just root range
    expect(container.firstChild).toMatchSnapshot();

    // With committed range
    dispatch(ProfileView.commitRange(0, 40));
    expect(container.firstChild).toMatchSnapshot();

    // With preview selection
    dispatch(
      ProfileView.updatePreviewSelection({
        hasSelection: true,
        isModifying: false,
        selectionStart: 10,
        selectionEnd: 10.1,
      })
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays the "Full Range" text as its first element', () => {
    const { getByText } = setup();
    expect(getByText('Full Range')).toBeTruthy();
  });

  it('renders the site hostname as its first element in the single tab view', () => {
    const { dispatch, container } = setup();
    dispatch(
      ReceiveProfile.changeTimelineTrackOrganization({
        type: 'active-tab',
        browsingContextID,
      })
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('displays the site hostname as its first element in the single tab view', () => {
    const { dispatch, queryByText, getByText } = setup();
    dispatch(
      ReceiveProfile.changeTimelineTrackOrganization({
        type: 'active-tab',
        browsingContextID,
      })
    );
    expect(queryByText('Full Range')).toBeFalsy();
    // Using regexp because searching for a partial text.
    expect(getByText(/developer\.mozilla\.org/)).toBeTruthy();
  });
});
