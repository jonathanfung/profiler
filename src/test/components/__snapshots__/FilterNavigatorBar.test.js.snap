// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/ProfileFilterNavigator renders ProfileFilterNavigator properly 1`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
    data-index="0"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      Full Range
    </span>
  </li>
</ol>
`;

exports[`app/ProfileFilterNavigator renders ProfileFilterNavigator properly 2`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarBeforeSelectedItem"
    data-index="0"
  >
    <button
      class="filterNavigatorBarItemContent"
      type="button"
    >
      Full Range
    </button>
  </li>
  <li
    class="filterNavigatorBarItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem filterNavigatorBarTransition-enter filterNavigatorBarTransition-enter-active"
    data-index="1"
    title="40ms"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      40ms
    </span>
  </li>
</ol>
`;

exports[`app/ProfileFilterNavigator renders ProfileFilterNavigator properly 3`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarBeforeSelectedItem"
    data-index="0"
  >
    <button
      class="filterNavigatorBarItemContent"
      type="button"
    >
      Full Range
    </button>
  </li>
  <li
    class="filterNavigatorBarItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem filterNavigatorBarTransition-enter filterNavigatorBarTransition-enter-active"
    data-index="1"
    title="40ms"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      40ms
    </span>
  </li>
  <li
    class="filterNavigatorBarItem filterNavigatorBarLeafItem filterNavigatorBarUncommittedItem filterNavigatorBarUncommittedTransition-enter filterNavigatorBarUncommittedTransition-enter-active"
    title="100μs"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      100μs
    </span>
  </li>
</ol>
`;

exports[`app/ProfileFilterNavigator renders the site hostname as its first element in the single tab view 1`] = `
<ol
  class="filterNavigatorBar profileFilterNavigator"
>
  <li
    class="filterNavigatorBarItem filterNavigatorBarRootItem filterNavigatorBarSelectedItem filterNavigatorBarLeafItem"
    data-index="0"
    title="https://developer.mozilla.org"
  >
    <span
      class="filterNavigatorBarItemContent"
    >
      <div
        class="nodeIcon "
      />
      developer.mozilla.org
       (
      51ms
      )
    </span>
  </li>
</ol>
`;
